storage/innobase/dict/dict0stats.cc

```diff
@@ -1372,6 +1382,88 @@ static dberr_t btr_pcur_open_level(btr_pcur_t *pcur, ulint level, mtr_t *mtr,
 }
 
 
+/** Statistics collected for a single B-tree level
+during index analysis. Used to track distinct key values,
+record counts, and page information when analyzing index
+levels for the query optimizer's statistics. */
+struct IndexLevelStats
+{
+  /** Level number */
+  uint16_t level;
+  /** Whether nulls are considered unequal for statistics */
+  bool nulls_unequal;
+  /** Array for number of distinct keys */
+  uint64_t *n_diff;
+  /** Array of number of non-null keys for all prefixes */
+  uint64_t *n_non_null;
+  /** Boundaries of groups of distinct keys */
+  boundaries_t *bounds;
+  /** Total number of records */
+  uint64_t n_recs;
+  /** Total number of pages */
+  uint32_t n_pages;
+
+  IndexLevelStats(unsigned stats_method, dict_index_t *index) :
+    level(0), nulls_unequal(stats_method > SRV_STATS_NULLS_EQUAL)
+  {
+    n_diff= index->stat_n_diff_key_vals;
+    memset(n_diff, 0x0, index->n_uniq * sizeof(*n_diff));
+    if (stats_method == SRV_STATS_NULLS_IGNORED)
+    {
+      n_non_null= index->stat_n_non_null_key_vals;
+      memset(n_non_null, 0x0, index->n_uniq * sizeof(*n_non_null));
+    }
+    else n_non_null= nullptr;
+    n_recs= 0;
+    n_pages= 0;
+    bounds= nullptr;
+  }
+
+  IndexLevelStats(uint16_t level_, unsigned stats_method,
+                  dict_index_t *index) :
+    level(level_), nulls_unequal(stats_method > SRV_STATS_NULLS_EQUAL)
+  {
+    n_diff=
+      static_cast<uint64_t*>(ut_zalloc(index->n_uniq * sizeof(*n_diff),
+                             mem_key_dict_stats_n_diff_on_level));
+    if (stats_method == SRV_STATS_NULLS_IGNORED)
+      n_non_null=
+	static_cast<uint64_t*>(ut_zalloc(index->n_uniq * sizeof(*n_non_null),
+                               mem_key_dict_stats_n_diff_on_level));
+    else n_non_null= nullptr;
+    n_recs= 1;
+    n_pages= 0;
+    bounds= UT_NEW_ARRAY_NOKEY(boundaries_t, index->n_uniq);
+  }
+
+  ~IndexLevelStats()
+  {
+    if (level)
+    {
+      UT_DELETE_ARRAY(bounds);
+      ut_free(n_diff);
+      ut_free(n_non_null);
+    }
+  }
+};
+
+/** Statistics collected for a leaf level during index
+analysis. Used to calculate the number of different key values,
+number of blob, number of non-null values in an index
+when looking at the first n_prefix columns. */
+struct LeafPageStats
+{
+  const uint8_t n_prefix;
+  const bool nulls_unequal;
+  const bool calc_non_null;
+  uint32_t n_blob= 0;
+  uint64_t n_diff= 0;
+  uint64_t n_non_null= 0;
```

Comment:
The data members are not documented. In particular, the `n_blob` could be misleading. Is it about BLOB count, or an approximate count of individual BLOB pages? Which function or expression will it be initialized from?